src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperCompletionService.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    [Export(typeof(TagHelperCompletionService))]
+    internal class DefaultTagHelperCompletionService : TagHelperCompletionService
+    {
+        private readonly TagHelperFactsService _tagHelperFactsService;
+
+        [ImportingConstructor]
+        public DefaultTagHelperCompletionService(TagHelperFactsService tagHelperFactsService)
+        {
+            _tagHelperFactsService = tagHelperFactsService;
+        }
+
+        public override ElementCompletionResult GetElementCompletions(ElementCompletionContext completionContext)
+        {
+            if (completionContext == null)
+            {
+                throw new ArgumentNullException(nameof(completionContext));
+            }
+
+            var elementCompletions = new Dictionary<string, IEnumerable<TagHelperDescriptor>>(StringComparer.OrdinalIgnoreCase);
+
+            AddAllowedChildrenCompletions(completionContext, elementCompletions);
+
+            if (elementCompletions.Count > 0)
+            {
+                // If there are allowed children, those are the only possible completions.
```

Comment:
I would say something like "if this element is already a taghelper, and that taghelper only allows certain children....."